test/cmd/newapp.sh

```diff
@@ -470,5 +470,18 @@ os::cmd::expect_success_and_not_text 'oc new-app https://github.com/openshift/ru
 # We permit running new-app against a remote URL which returns a template
 os::cmd::expect_success 'oc new-app https://raw.githubusercontent.com/openshift/origin/master/examples/wordpress/template/wordpress-mysql.json --dry-run'
 
+# new-app different syntax for new-app functionality
+os::cmd::expect_success 'oc import-image openshift/ruby-20-centos7:latest --confirm'
+rm -rf ./test/testdata/testapp
+git clone https://github.com/openshift/ruby-hello-world.git ./test/testdata/testapp
+os::cmd::expect_success 'oc new-app ruby-20-centos7:latest~https://github.com/openshift/ruby-hello-world.git --dry-run'
+os::cmd::expect_success 'oc new-app ruby-20-centos7:latest~./test/testdata/testapp --dry-run'
+os::cmd::expect_success 'oc new-app -i ruby-20-centos7:latest https://github.com/openshift/ruby-hello-world.git --dry-run'
+os::cmd::expect_success 'oc new-app -i ruby-20-centos7:latest ./test/testdata/testapp --dry-run'
+os::cmd::expect_success 'oc new-app ruby-20-centos7:latest --code https://github.com/openshift/ruby-hello-world.git --dry-run'
+os::cmd::expect_success 'oc new-app ruby-20-centos7:latest --code ./test/testdata/testapp --dry-run'
+os::cmd::expect_success 'oc new-app -i ruby-20-centos7:latest --code https://github.com/openshift/ruby-hello-world.git --dry-run'
+os::cmd::expect_success 'oc new-app -i ruby-20-centos7:latest --code ./test/testdata/testapp --dry-run'
```

Comment:
in all these cases since we are supplying the builder image explicitly, we should expect that source detection will not be performed on the source argument.  Add something to the test that confirms that:

1) use a non-ruby image as the builder image and confirm the resulting build uses the non-ruby builder image (even though the source is ruby code)
2) run new-app with a higher log level and check the output for indications that source detection is being performed (it should not be).

Reply:
For point 2 code detection still happens, but the image is being used from the flags and not detection.
`oc new-app php-55-centos7:latest --code ./app-examples/ruby --dry-run --loglevel=4`
result: php

`oc new-app --code ./app-examples/ruby --dry-run --loglevel=4`
Result: ruby
Will do test to confirm. 

Problem is that code detection is doing a little bit more than code detection so it needs to be executed[1]. We might want to split it. But I would not want to touch this one here as the logic to use flags values was already there.  

[1]
```
I1215 20:51:28.617427   28984 resolve.go:254] checkig source code type for [./app-examples/ruby]
I1215 20:51:28.618698   28984 repository.go:385] Executing git config --get-regexp ^remote\..*\.url$
I1215 20:51:28.622574   28984 repository.go:385] Executing git symbolic-ref -q --short HEAD
I1215 20:51:28.625057   28984 sourcelookup.go:313] Checking if https://github.com/openshift/ruby-hello-world requires authentication
I1215 20:51:28.625074   28984 repository.go:385] Executing git ls-remote --heads https://github.com/openshift/ruby-hello-world
```

Reply:
Found this strange behavior which I think is not as expected (its on 3.7 too and was not introduced by these code changes). It would be nice you could validate. Syntax are similar - one uses dicovery for images, other overwrites. BUT strategies in the result are different. 
```
os::cmd::expect_success 'oc new-app --code ./test/testdata/testapp --name test'
os::cmd::expect_success_and_text 'oc get bc test --template={{.spec.strategy.dockerStrategy.from.name}}' 'ruby-22-centos7:latest'
os::cmd::expect_success 'oc new-app -i php-55-centos7:latest --code ./test/testdata/testapp --name test2'
os::cmd::expect_success_and_text 'oc get bc test2 --template={{.spec.strategy.sourceStrategy.from.name}}' 'php-55-centos7:latest'
```

Reply:
what is in /test/testdata/testapp in the first test?  if it contains a Dockerfile with `FROM ruby-22-centos7:latest` then i'd say it's behaving as expected.

Reply:
you right :)

Reply:
> in all these cases since we are supplying the builder image explicitly, we should expect that source detection will not be performed on the source argument. Add something to the test that confirms that:
>
>    use a non-ruby image as the builder image and confirm the resulting build uses the non-ruby builder image 
> (even though the source is ruby code)

I don't see that this test was added for the various permutations of arguments (`image~source`, `image source`, `-i image --source source`, etc).

Reply:
This was covered by 2 tests, but exteded with few more few more tests:
```
...
oc new-app php-55-centos7:latest --code https://github.com/openshift/ruby-hello-world.git --name test5
oc get bc test5 --template={{.spec.strategy.sourceStrategy.from.name}}
#result is php builder, despite the fact that code is ruby.
....
```

This covers all your cases. The only tests we don't have is how it behaves without git. This is due our testing infra limitation :/